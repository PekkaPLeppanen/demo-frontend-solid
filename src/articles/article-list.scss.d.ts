declare const styles: {
	readonly 'articleList': string;
};
export = styles;
