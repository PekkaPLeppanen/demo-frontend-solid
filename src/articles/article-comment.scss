.article-comment {
  border-top: 1px solid #dddddd;
  padding: 0.5em;

  &__email {
    color: #666;
    font-size: 0.75em;
  }

  &__body {
    padding: 0 0.5em;
    font-size: 0.875em;
  }
}
