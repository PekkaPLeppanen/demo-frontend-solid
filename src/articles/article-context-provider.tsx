import { createContext, createSignal, type FlowProps, type JSX, type Signal, useContext } from 'solid-js'
import { type IArticle } from '@/resources/article-provider'

type IArticleContext = Signal<IArticle | undefined>

/**
 * What is returned here can be accessed in the child components of the ArticleProvider through useArticle()
 */
const makeArticleContext = (article?: IArticle): IArticleContext => {
  return createSignal(article)
}

const ArticleContext = createContext<IArticleContext>(makeArticleContext())

/**
 * Creates a context provider component, where all the children have access to the ArticleContext
 *
 * @see https://www.solidjs.com/docs/latest/api#createcontext
 *
 * @param props
 * @constructor
 */
export function ArticleContextProvider (props: FlowProps): JSX.Element {
  return (
    <ArticleContext.Provider value={makeArticleContext()}>
      {props.children}
    </ArticleContext.Provider>
  )
}

/**
 * Usage: const [isMenuOpen, {openMenu, closeMenu}] = useArticle();
 */
export function useArticle (): IArticleContext {
  return useContext(ArticleContext)
}
