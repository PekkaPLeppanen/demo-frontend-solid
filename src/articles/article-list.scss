.article-list {
  list-style: none;

  &__item {
    text-transform: capitalize;
    padding: 1em;
    margin-top: 1em;
    border: 1px solid #ddd;
    border-radius: 5px;
  }
}
