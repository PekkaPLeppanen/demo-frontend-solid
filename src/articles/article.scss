.article {
  padding: 0 0.5em;

  &__title {}

  &__body {
    &:after {
      content: '.'
    }
  }
}
