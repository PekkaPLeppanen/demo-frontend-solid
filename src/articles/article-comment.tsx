import {Component} from 'solid-js';
import {IArticleComment} from '@/resources/article-provider';
import '@/articles/article-comment.scss';

export type ArticleCommentProps = {
	comment: IArticleComment;
}

export const ArticleComment: Component<ArticleCommentProps> = (props) => {
	return (
		<div class="article-comment">
			<div class="article-comment__email">{props.comment.email}</div>
			<div class="article-comment__body">{props.comment.body}</div>
		</div>
	);
};
