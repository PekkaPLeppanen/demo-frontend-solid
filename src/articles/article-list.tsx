import { type Component, createResource, For } from 'solid-js'
import { getAllArticles, type IArticle } from '@/resources/article-provider'
import { Suspense } from 'solid-js/web'
import '@/articles/article-list.scss'
import { A } from '@solidjs/router'
import { useArticle } from '@/articles/article-context-provider'

export const ArticleList: Component = () => {
  const [, selectArticle] = useArticle()
  const [articles] = createResource(getAllArticles)

  return (
    <>
      <h1>Latest topics</h1>
      <Suspense fallback={<p>Loading...</p>}>
        <ul class="article-list">
          <For each={articles()}>{(article: IArticle) =>
            <li class="article-list__item">
              <A href={`/article/${article.id}`} onClick={() => {
                selectArticle(article)
              }}>{article.title}</A>
            </li>
          }</For>
        </ul>
      </Suspense>
    </>

  )
}
