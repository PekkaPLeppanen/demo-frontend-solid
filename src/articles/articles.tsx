import {Component, createResource, For} from 'solid-js';
import {Article, getAllArticles} from '@/resources/article-provider';
import {Suspense} from 'solid-js/web';
import '@/articles/article-list.scss';

export const Articles: Component = () => {

	const [articles] = createResource(getAllArticles);

	return (
		<>
			<h1>Latest topics</h1>
			<Suspense fallback={<p>Loading...</p>}>
				<ul class="article-list">
					<For each={articles()}>{(article: Article) =>
						<li class="article-list__item">
							{article.title}
						</li>
					}</For>
				</ul>
			</Suspense>
		</>

	);
};
