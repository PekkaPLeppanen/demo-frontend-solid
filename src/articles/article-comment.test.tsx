import { render } from '@solidjs/testing-library'
import { ArticleComment } from '@/articles/article-comment'
import { expect } from 'vitest'
import { articleCommentsResponseMock } from '@/resources/__mocks__/articles'

describe('<ArticleComment />', () => {
  test('render', async () => {
    const articleComment = articleCommentsResponseMock[0]
    const { container, getByText, unmount } = render(() => <ArticleComment comment={articleComment}/>)
    expect(container.outerHTML).toMatchSnapshot()
    expect(getByText(articleComment.body)).toBeInTheDocument()
    expect(getByText(articleComment.email)).toBeInTheDocument()
    unmount()
  })
})
