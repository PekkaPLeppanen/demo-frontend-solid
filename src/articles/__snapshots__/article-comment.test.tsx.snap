// Vitest Snapshot v1

exports[`<ArticleComment /> > render 1`] = `"<div><div class=\\"article-comment\\"><div class=\\"article-comment__email\\">Eliseo@gardner.biz</div><div class=\\"article-comment__body\\">Laudantium enim quasi est quidem magnam voluptate ipsam eos tempora quo necessitatibus dolor quam autem quasi reiciendis et nam sapiente accusantium</div></div></div>"`;
